File: backend/models/game.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type GameStatus string

const (
	StatusInProgress GameStatus = "in_progress"
	StatusWon        GameStatus = "won"
	StatusLost       GameStatus = "lost"
)

type GuessResult struct {
	Guess     string    `bson:"guess"     json:"guess"`
	Result    []string  `bson:"result"    json:"result"`
	Timestamp time.Time `bson:"timestamp" json:"timestamp"`
}

type Game struct {
	ID          primitive.ObjectID `bson:"_id,omitempty"        json:"id"`
	UserID      primitive.ObjectID `bson:"userId"               json:"userId"`
	Word        string             `bson:"word"                 json:"-"` // never sent to frontend
	Guesses     []GuessResult      `bson:"guesses"              json:"guesses"`
	Status      GameStatus         `bson:"status"               json:"status"`
	StartedAt   time.Time          `bson:"startedAt"            json:"startedAt"`
	CompletedAt *time.Time         `bson:"completedAt,omitempty" json:"completedAt,omitempty"`
}

File: game/lib/models/Game.ts
import mongoose, { Schema, Document, Types } from 'mongoose';

export type GameStatus = 'in_progress' | 'won' | 'lost';

export interface GuessEntry {
  guess: string;
  result: string[]; // ["+" | "x" | "-", ...]
  timestamp: Date;
}

export interface IGame extends Document {
  userId: Types.ObjectId;
  word: string; // never sent to the client
  guesses: GuessEntry[];
  status: GameStatus;
  startedAt: Date;
  completedAt?: Date;
}

const GuessEntrySchema = new Schema<GuessEntry>(
  { guess: String, result: [String], timestamp: Date },
  { _id: false }
);

const GameSchema = new Schema<IGame>({
  userId:      { type: Schema.Types.ObjectId, ref: 'User', required: true },
  word:        { type: String, required: true },
  guesses:     { type: [GuessEntrySchema], default: [] },
  status:      { type: String, enum: ['in_progress', 'won', 'lost'], default: 'in_progress' },
  startedAt:   { type: Date, default: Date.now },
  completedAt: { type: Date },
});

export default (mongoose.models.Game as mongoose.Model<IGame>) ||
  mongoose.model<IGame>('Game', GameSchema);
